### PR TITLE
fix(cubemaster): clean job rows after snapshot delete

### PR DESCRIPTION
**Pull request by xiaojunxiang2023**
## Summary

Fixes a bug where deleted snapshots still appeared in template list APIs until a second delete.

After the first `DELETE /templates/{snapshotID}`:

- Single-snapshot lookup correctly returned not found
- `GET /templates` and `cubemastercli tpl ls` still listed the deleted snapshot
- A second delete was required to remove it from the list
image

## Root cause

Snapshot delete (`runSnapshotDeleteJob`) removed `template_definition` metadata but did not remove related `template_image_job` rows.

`ListTemplates` appends orphan job records when no definition exists:

```go
for _, job := range jobs {
    if _, ok := seen[job.TemplateID]; ok {
        continue
    }
    out = append(out, templateInfoFromJob(&job))
}
```

So the deleted snapshot was resurrected in list responses via job fallback entries.

On the second delete, `has_snapshot` returned false (definition gone, job fallback has no `Kind`), routing the request through template delete (`cleanupTemplateJobs`) instead of snapshot delete.

## Fix

- Call `cleanupTemplateJobs` after successful snapshot metadata cleanup, matching template delete behavior
- Return terminal READY job info from `executeSnapshotDeleteJob` without re-querying job rows that were just deleted

## Deploy And Verify
image

### CODE REVIEW

File: CubeMaster/pkg/templatecenter/snapshot_ops_test.go [hunk 1]
```diff
@@ -608,3 +608,91 @@ func TestDeleteSnapshotDoesNotBlockWhenRuntimeRefsExist(t *testing.T) {
...
+	patches.ApplyFunc(discoverTemplateCleanupTargets, func(ctx context.Context, templateID, instanceType string) (*templateCleanupTargets, error) {
+		return &templateCleanupTargets{}, nil
+	})
+	patches.ApplyFunc(snapshotDeleteLocators, func(targets *templateCleanupTargets) ([]templateCleanupLocator, error) {
+		return nil, nil
+	})
+	patches.ApplyFunc(runReplicaCleanup, func(ctx context.Context, templateID string, locators []templateCleanupLocator) error {
+		return nil
+	})
+	patches.ApplyFunc(deleteSnapshotMetadataOnly, func(ctx context.Context, snapshotID string) error {
```
Comment (fslongjin) [inline CubeMaster/pkg/templatecenter/snapshot_ops_test.go]: This test is risky for CI under the default Go compiler settings. `deleteSnapshotMetadataOnly` is just a tiny wrapper, so the compiler may inline it; when that happens, `gomonkey.ApplyFunc(deleteSnapshotMetadataOnly, ...)` does not intercept the call and the test still reaches the real `cleanupTemplateMetadata`. Since the test sets `store.db` to an empty `&gorm.DB{}`, that real DB path can panic.

I can reproduce this locally: `go test ./pkg/templatecenter` panics in `TestRunSnapshotDeleteJobCleansTemplateJobs`, while `go test -gcflags=all=-l ./pkg/templatecenter` passes after disabling inlining. Consider avoiding a patch on this tiny wrapper, for example by introducing a replaceable function variable or by using a real test DB to verify the job cleanup behavior.

Comment (xiaojunxiang2023) [reply CubeMaster/pkg/templatecenter/snapshot_ops_test.go]: Thanks for catching this — reproduced the panic locally as well.

Addressed in the latest push:

1. `runSnapshotDeleteJob` now calls the replaceable hooks `runMetadataCleanup` and `runTemplateJobCleanup` instead of the inlined wrappers `deleteSnapshotMetadataOnly` / `cleanupTemplateJobs`.
2. `TestRunSnapshotDeleteJobCleansTemplateJobs` follows the same pattern as `delete_test.go`: swap the function variables in `t.Cleanup` instead of patching tiny wrappers with gomonkey. The test no longer needs a dummy `store.db`.

Verified with default compiler settings:
`go test ./pkg/templatecenter -run TestRunSnapshotDeleteJobCleansTemplateJobs`

File: CubeMaster/pkg/templatecenter/snapshot_ops.go [hunk 2]
```diff
@@ -708,16 +708,13 @@ func runSnapshotDeleteJob(ctx context.Context, jobID, snapshotID string) error {
...
-		"status":      JobStatusReady,
-		"phase":       JobPhaseReady,
-		"progress":    100,
-		"result_json": `{"deleted":true}`,
-	}); err != nil {
+	// Mirror template delete: drop job rows so ListTemplates does not
+	// resurrect the snapshot from orphan job fallback entries.
+	if err := runTemplateJobCleanup(ctx, snapshotID); err != nil {
 		return err
 	}
```
[the remark this answers is not in the document]
Comment (xiaojunxiang2023) [reply CubeMaster/pkg/templatecenter/snapshot_ops.go]: Intentional — this matches `deleteTemplateWithTargets` in delete.go, which also returns the raw error from `runTemplateJobCleanup` after metadata is already removed (L119-121).

At this point replicas and definition are gone, so the snapshot is effectively deleted. Calling `failSnapshotDeleteJob` would attempt to UPDATE a definition row that no longer exists (silent no-op) and mark the job FAILED, leaving a phantom orphan entry in ListTemplates — the exact bug this PR fixes.

Returning the error directly is consistent with the existing template delete path.

File: CubeMaster/pkg/templatecenter/job_dto_test.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+		RedoScope:  []string{"phase-a"},
+	}
+	dst := cloneTemplateImageJobInfo(src)
+	if dst == nil || dst == src {
+		t.Fatal("expected independent clone")
+	}
+	dst.RedoScope[0] = "changed"
+	if src.RedoScope[0] == "changed" {
+		t.Fatal("RedoScope should be deep-copied")
+	}
```
[the remark this answers is not in the document]
Comment (xiaojunxiang2023) [reply CubeMaster/pkg/templatecenter/job_dto_test.go:31]: Acknowledged. The clone helper is a straightforward deep-copy for the in-memory READY response; the critical regression coverage for this PR is:

1. `runTemplateJobCleanup` is called after metadata delete (TestRunSnapshotDeleteJobCleansTemplateJobs)
2. `executeSnapshotDeleteJob` does not re-query deleted job rows (TestExecuteSnapshotDeleteJobReturnsReadyWithoutJobLookup)

Expanding clone tests for Artifact/Template branches is low value for this bugfix and increases PR scope. Happy to add broader clone coverage in a follow-up if maintainers prefer.

File: CubeMaster/pkg/templatecenter/snapshot_ops.go [hunk 4]
```diff
@@ -708,16 +708,13 @@ func runSnapshotDeleteJob(ctx context.Context, jobID, snapshotID string) error {
...
 	invalidateTemplateCaches(snapshotID)
-	if err := updateTemplateImageJob(ctx, jobID, map[string]any{
-		"status":      JobStatusReady,
-		"phase":       JobPhaseReady,
-		"progress":    100,
-		"result_json": `{"deleted":true}`,
-	}); err != nil {
+	// Mirror template delete: drop all job rows so ListTemplates does not
+	// resurrect the snapshot from orphan job fallback entries.
+	if err := runTemplateJobCleanup(ctx, snapshotID); err != nil {
```
[the remark this answers is not in the document]
Comment (xiaojunxiang2023) [reply CubeMaster/pkg/templatecenter/snapshot_ops.go:717]: Same behavior as `deleteTemplateWithTargets` in delete.go (L119-121): after metadata is gone, job cleanup failure returns the error directly. The happy path deletes all job rows and fixes the ListTemplates ghost-entry bug — that's what this PR targets. Post-metadata cleanup failure is a pre-existing edge case, not introduced here. Out of scope for this minimal bugfix.

File: CubeMaster/pkg/templatecenter/job_dto_test.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+package templatecenter
+
+import (
+	"testing"
+
+	sandboxtypes "github.com/tencentcloud/CubeSandbox/CubeMaster/pkg/service/sandbox/types"
+)
+
+func TestCloneTemplateImageJobInfo(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (xiaojunxiang2023) [reply CubeMaster/pkg/templatecenter/job_dto_test.go:13]: Acknowledged. For this bugfix the important tests are: (1) `runTemplateJobCleanup` is called after metadata delete, (2) no re-query of deleted job rows. Artifact/Template clone branches are not on the delete hot path — deferring to keep the PR minimal.

File: CubeMaster/pkg/templatecenter/job_dto_test.go [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+	src := &sandboxtypes.TemplateImageJobInfo{
+		JobID:      "job-1",
+		TemplateID: "tpl-1",
+		Status:     JobStatusPending,
+		RedoScope:  []string{"phase-a"},
+	}
+	dst := cloneTemplateImageJobInfo(src)
+	if dst == nil || dst == src {
+		t.Fatal("expected independent clone")
+	}
```
[the remark this answers is not in the document]
Comment (xiaojunxiang2023) [reply CubeMaster/pkg/templatecenter/job_dto_test.go:27]: Already addressed in an earlier reply — out of scope for this minimal bugfix. The critical tests here are job cleanup invocation and no re-query after delete. Happy to expand clone coverage in a follow-up if maintainers want it.

File: CubeMaster/pkg/templatecenter/snapshot_ops_test.go [hunk 7]
```diff
@@ -608,3 +608,96 @@ func TestDeleteSnapshotDoesNotBlockWhenRuntimeRefsExist(t *testing.T) {
...
+	})
+	patches.ApplyFunc(invalidateTemplateCaches, func(templateID string) {})
+	runReplicaCleanup = func(ctx context.Context, templateID string, locators []templateCleanupLocator) error {
+		return nil
+	}
+	runMetadataCleanup = func(ctx context.Context, templateID string) error {
+		return nil
+	}
+	runTemplateJobCleanup = func(ctx context.Context, templateID string) error {
+		if templateID != "snap-del" {
```
[the remark this answers is not in the document]
Comment (xiaojunxiang2023) [reply CubeMaster/pkg/templatecenter/snapshot_ops_test.go:643]: Acknowledged. This matches delete.go behavior: after metadata delete, job cleanup failure returns the raw error. Testing that failure branch is a pre-existing edge case shared with template delete, not introduced by this PR. Keeping the test suite focused on the happy-path regression this PR fixes.
